File: 0325-maximum-size-subarray-sum-equals-k/0325-maximum-size-subarray-sum-equals-k.py
class Solution:
    def maxSubArrayLen(self, nums: List[int], k: int) -> int:
        psum = max_len = 0
        indices = {0: -1}

        for i, num in enumerate(nums):
            psum += num

            if psum - k in indices:
                max_len = max(max_len, i - indices[psum - k])
            
            if psum not in indices:
                indices[psum] = i

        return max_len

File: 0325-maximum-size-subarray-sum-equals-k/0325-maximum-size-subarray-sum-equals-k.cpp
class Solution {
public:
    int maxSubArrayLen(vector<int>& nums, int k) {
        int n = nums.size(), maxLen = 0;
        long psum = 0;
        unordered_map<long, int> indices{{0, -1}};

        for (int i = 0; i < n; ++i){
            psum += nums[i];

            if (indices.count(psum - k)) maxLen = max(maxLen, i - indices[psum - k]);

            if (!indices.count(psum)) indices[psum] = i;
        }

        return maxLen;
    }
};